### ISSUE TITLE
Assignment of immutable record into mutable record variable does not produce error

### ISSUE THREAD

**Issue by MrWolfZ** [microsoft/TypeScript#52225]
# Bug Report

### 🔎 Search Terms

readonly

### 🕗 Version & Regression Information

I saw this behavior on TypeScript 4.9.4.

### ⏯ Playground Link

[bug workbench with relevant code](https://www.typescriptlang.org/dev/bug-workbench/?ts=4.9.4#code/MYewdgzgLgBAlgWwQVygQwEYBsCmAlHUAJwBMAuGAtE8LATwB4DiSHoi4wBzAGhmTABrMCADuYAHwSYAXhgBvAL4BuALAAoRCnTZ8hEKQB0AMxAhZMAOQY0RS8pgB6RzDQQYOAB4AHQlBwkfN5EICTIwDjuOEQhRBqgkLDamLjMBuSU+qRsUBzcfALCYpIWWqgpeiwOzjBQABZw7hB1IMhYJB4xBnwYqDA0kTAiUBrJumlGpuZy1rb2Ti5uHj5+AXwinbEaQA)

### 💻 Code

```ts
const immutableRecord: Readonly<Record<string, unknown>> = {};
immutableRecord.foo = 'bar'; // as expected, produces error
const mutableRecord: Record<string, unknown> = immutableRecord; // this should error, but does not
mutableRecord.foo = 'bar'; // as expected, no error
```

### 🙁 Actual behavior

I noticed that a readonly record can be assigned to a mutable record without a compile error. I assume this is because everything is assignable to the empty object (as described [here](https://github.com/Microsoft/TypeScript/wiki/FAQ#why-do-these-empty-classes-behave-strangely)) but it is very unintuitive, since we are not actively creating a class or type with no properties. If you ask any developer, they would assume this code to produce an error.

### 🙂 Expected behavior

Optimally, the assignment of an immutable record into a mutable record variable should not be allowed.

**Comment by fatcerberus** [comment]
This is not specific to `Record`.  In general `{ readonly foo: T }` is assignable to `{ foo: T }`.  It doesn’t really mean “immutable”, but rather a much weaker guarantee - only that it can’t be mutated *through the `readonly` reference*.

**Comment by MartinJohns** [comment]
Hence why I always try to tell people: readonly does not mean immutable. It only means readonly via that interface.

**Comment by Fryuni** [comment]
The behavior is inconsistent:
```ts
const immutableRecord: Readonly<Record<string, unknown>> = {};
// Doesn't fail
const mutableRecord: Record<string, unknown> = immutableRecord;

const immutableArray: ReadonlyArray<string> = [];
// This breaks, as it should
// The type 'readonly string[]' is 'readonly' and cannot be assigned to the mutable type 'string[]'
const mutableArray: Array<string> = immutableArray;
```

**Comment by fatcerberus** [comment]
It’s consistent w.r.t. to the structural typing rules: the readonly array is not assignable only as a side effect of missing methods, not directly due to its readonly-ness.

**Comment by Fryuni** [comment]
> It’s consistent w.r.t. to the structural typing rules: the readonly array is not assignable only as a side effect of missing methods, not directly due to its readonly-ness (despite what the error message says)

Aka, it is inconsistent. From the user's perspective, it doesn't really matter if the part that works as expected (the array) works because someone made it work or because of a happy coincidence.

**Comment by fatcerberus** [comment]
It’s consistent by the assignability rules relative to other types.  The fact that the `readonly` keyword is being used for two separate concepts is unfortunate, but overloading of keywords is hardly unique to TypeScript.

**Comment by fatcerberus** [comment]
fwiw the line between “works by happy coincidence” and “someone made it work this way” tend to be blurred in a structural type system because the compiler only cares about the contents of your type and not what keywords were used to construct it.  Higher-order expectations don’t really translate well in that environment.

**Comment by MrWolfZ** [comment]
well, you learn something new every day. this doesn't seem like it will be changed, so I'll close this issue
